### DISCUSSION TITLE
[Proposal]: Use CompareTo to resolve missing comparison operators

### DISCUSSION THREAD

**Discussion by VBAndCs** [dotnet/csharplang#4119] [Language Ideas]
Given this class that implements the IComparable:
```C#
class Test : IComparable< int>{

    public int Value;
        public int CompareTo(int other) {
            return Value.CompareTo(other);
       }
    }
```

Why can't c# use to the CompareTo method to resolve the comparison operators?
```c#
var f = new Test() { Value = 1};
Console.WriteLine( f == 1);
Console.WriteLine( f != 1);
Console.WriteLine(f > 1);
Console.WriteLine(f >= 1);
Console.WriteLine(f < 1);
Console.WriteLine(f <= 1);
```

which can be loqered to:
var f = new Test() { Value = 1};
Console.WriteLine(f.CompareTo(1) == 0);
Console.WriteLine(f.CompareTo(1) != 0);
Console.WriteLine(f.CompareTo(1) == 1);
Console.WriteLine(f.CompareTo(1) != -1);
Console.WriteLine(f.CompareTo(1) == -1);
Console.WriteLine(f.CompareTo(1) != 1);
```

**Comment by PathogenDavid** [comment]
Since this isn't a fully-specified language proposal, it should've been opened as a discussion.

Personally I'm not a fan, and I don't think this could be introduced without it being a breaking change. It seems like a good candidate for a source generator to me.

**Comment by VBAndCs** [comment]
> think this could be introduced without it being a breaking change

How? The code doesn't compile today, so it can't break anything.
And if there is an overloaded comparison operator, it will take precedence, so, ole code will run as usual.

**Comment by HaloFour** [comment]
@VBAndCs 

> The code doesn't compile today

It does for the equality operator.

**Comment by VBAndCs** [comment]
> It does for the equality operator.

No, it doesn't (.net core 3.1).

**Comment by PathogenDavid** [comment]
Yes it does. You'd be changing the meaning of `==` from reference equality to use `IComparable`.

See [this example](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEUCuA7AHwAEAmARgFgAoaogZgAJSGAVGAZwwZAYEkBhCAFsADgEMoY4ABsYAHjacAfNQDe1BpqaMAlni4A1MdJwwGqhgHMYGANwMAvtQ1b6rDhgAUergDdjpgCULlpaALxKDEYmZmEM/jG2IaEMyW4+DIKiEjAsEJ6KXBAYABYwUMFUKeGR0aYAdFnisHmexWVQ9XUwgUlUTjRUbsyFAOo6pShkZACc3HxNElKyCh7jk9MzKlTqVa66+lEBZhbWdo7Oe5puYxMlU7PehwlByaERRzEMcS8wfdWaNIHLiLFr5W4bWYMdrlSoA761Y6NYTNXL5GGdbq9S6hYZkJAMYAQCDSaEicpiDDQMJhAprO4POZiNDuTjre6bQlwgEfMTI7JgzzAQLfOIABn+WiBTHxhOJpIg5MkVKgAEJaRCOVDmayMOzGVy3jUGHzQWihSL1QwJdQBrQSHQ1GkSCRkrt4YUTd8GHgYAB3XWeMhisXYq4pT3Ab2+gOFIMhsPwoiPMSirmS6pGzSag1gaP+3X6zbx0MZiP0yFzAAm+djFa1MxLiYBycbeZpDCrzc0AwcQA) where `TestWith4119` is using `CompareTo` for `operator==` as you're requesting. They don't have the same result.

**Comment by VBAndCs** [comment]
@PathogenDavid
Still will not break anything. My suggestion takes place when there is no way to resolve the operator, so, the compiler starts looking if the object implements Icompareable as the last resort, so, I don't want to change any existing rules. My motivation here was this issue https://github.com/dotnet/roslyn/issues/49095. C# doesn't suffer with arithmetic operators, but still has an issue with comparisons.

**Comment by PathogenDavid** [comment]
Ah, I missed the part where you were implementing `IComparable<int>` instead of `IComparable<Test>` in your example.

I feel like if you're going to go that route it's just going to be this weird inconsistency where this only works when comparing reference types and value types or value types with other value types.

In that issue you've mentioned you're auto-generating these types. Why not just auto-generate your operator overloads too?

**Comment by VBAndCs** [comment]
> Why not just auto-generate your operator overloads too?

Recently. this seems to be the solution. But I wonder why can't the compiler deduce most of the operators if it has just enough info? In fact this doesn't apply only on CompareTo, but it is applicable also if the compiler has one of (`=` or `!=`) and on of (`<` or `>`), so it can use them to do the remain operators. And of course when inheriting base class operators they should be altered to deal with the child type, because it is useless now. The operator overloading needs a second look.
